src/ColinHDev/CPlot/provider/DataProvider.php

```diff
@@ -1291,4 +1306,255 @@ public function savePlotRate(Plot $plot, PlotRate $plotRate) : Generator {
+                        }
+
+                        // register player data
+                        yield from $this->updatePlayerData(
+                            $UUID->getBytes(), // doesn't matter what is input at this point. will overwrite on login
+                            $XUID,
+                            $playerName
+                        );
+                    }
+
+                    $playerData = yield $this->awaitPlayerDataByData(
+                        $UUID?->getBytes(),
+                        $XUID,
+                        $playerName
+                    );
+                    if (!($playerData instanceof PlayerData))
+                        continue;
+
+                    $senderData = new PlotPlayer($playerData, PlotPlayer::STATE_HELPER);
+                    $plot->addPlotPlayer($senderData);
+                    yield from DataProvider::getInstance()->savePlotPlayer($plot, $senderData);
+                }
+
+                // load denied with priority over helpers
+                foreach($record["denied"] as $playerName) {
+                    // validate offline player data
+                    $UUID = $XUID = null;
+                    $offlineData = Server::getInstance()->getOfflinePlayerData($playerName);
+                    if($offlineData !== null) {
+                        $UUID = $XUID = $offlineData->getString("LastKnownXUID", "");
+                        if($UUID === "") {
+                            $skinTag = $offlineData->getCompoundTag("Skin");
+                            $skinData = $skinTag->getByteArray("Data");
+                            $UUID = Uuid::uuid3(Uuid::NIL, ((string)Entity::nextRuntimeId()) . $skinData . $playerName);
+                        }else{
+                            $UUID = Uuid::fromInteger($UUID);
+                        }
+
+                        // register player data
+                        yield from $this->updatePlayerData(
+                            $UUID->getBytes(), // doesn't matter what is input at this point. will overwrite on login
+                            $XUID,
+                            $playerName
+                        );
+                    }
+
+                    $playerData = yield $this->awaitPlayerDataByData(
+                        $UUID?->getBytes(),
+                        $XUID,
+                        $playerName
+                    );
+                    if (!($playerData instanceof PlayerData))
+                        continue;
+
+                    $senderData = new PlotPlayer($playerData, PlotPlayer::STATE_DENIED);
+                    $plot->addPlotPlayer($senderData);
+                    yield from DataProvider::getInstance()->savePlotPlayer($plot, $senderData);
+                }
+
+                //load common flags
+                /** @var BaseAttribute<bool> | null $flag */
+                $flag = FlagManager::getInstance()->getFlagByID("pvp")->newInstance($record["pvp"]);
```

Comment:
My bad..
Since this branch wasn't updated, this still uses the old attribute system.
I merged `development` into `myplot-support` yesterday. So you could update it here.